core/imds/src/main/java/software/amazon/awssdk/imds/internal/TokenResponseHandler.java

```diff
@@ -16,55 +16,45 @@
 
 import static software.amazon.awssdk.imds.internal.RequestMarshaller.EC2_METADATA_TOKEN_TTL_HEADER;
-import static software.amazon.awssdk.imds.internal.StringResponseHandler.uncheckedInputStreamToUtf8;
 
 import java.time.Duration;
 import java.util.concurrent.CompletableFuture;
 import software.amazon.awssdk.annotations.SdkInternalApi;
-import software.amazon.awssdk.core.exception.RetryableException;
-import software.amazon.awssdk.core.exception.SdkClientException;
 import software.amazon.awssdk.core.http.HttpResponseHandler;
 import software.amazon.awssdk.core.interceptor.ExecutionAttributes;
-import software.amazon.awssdk.http.AbortableInputStream;
-import software.amazon.awssdk.http.HttpStatusFamily;
 import software.amazon.awssdk.http.SdkHttpFullResponse;
 import software.amazon.awssdk.utils.Logger;
 
 @SdkInternalApi
 public final class TokenResponseHandler implements HttpResponseHandler<Token> {
     private static final Logger log = Logger.loggerFor(TokenResponseHandler.class);
 
-    private CompletableFuture<Token> future;
     private final long ttlSeconds;
 
+    private final StringResponseHandler delegateHandler;
+
     public TokenResponseHandler(long ttlSeconds) {
         this.ttlSeconds = ttlSeconds;
+        this.delegateHandler = new StringResponseHandler();
     }
 
-    public void setFuture(CompletableFuture<Token> future) {
-        this.future = future;
+    public void setFuture(CompletableFuture<?> future) {
+        this.delegateHandler.setFuture(future);
     }
 
     @Override
     public Token handle(SdkHttpFullResponse response, ExecutionAttributes executionAttributes) throws Exception {
-        HttpStatusFamily statusCode = HttpStatusFamily.of(response.statusCode());
-        AbortableInputStream inputStream = response
-            .content().orElseThrow(() -> SdkClientException.create("Unexpected error: empty response content"));
-        String responseContent = uncheckedInputStreamToUtf8(inputStream);
-        if (statusCode.isOneOf(HttpStatusFamily.CLIENT_ERROR)) {
-            // non-retryable error
-            future.completeExceptionally(SdkClientException.builder()
-                                                           .message(responseContent)
-                                                           .build());
-        } else if (statusCode.isOneOf(HttpStatusFamily.SERVER_ERROR)) {
-            // retryable error
-            future.completeExceptionally(RetryableException.create(responseContent));
-        }
+        String tokenValue = delegateHandler.handle(response, executionAttributes);
         Duration ttl = response.firstMatchingHeader(EC2_METADATA_TOKEN_TTL_HEADER)
                                .map(Long::parseLong)
                                .map(Duration::ofSeconds)
-                               .orElseGet(() -> Duration.ofSeconds(ttlSeconds));
-        return new Token(responseContent, ttl);
+                               .orElseGet(() -> {
```

Comment:
Is it possible that the response will not have TTL?

Reply:
The server should always return a response with the ttl header. I was hesitating between this and throwing an exception if none were found.

Reply:
Can we confirm with the SEP author to see what the right behavior it should be?

Reply:
Talked to Matt, he pointed me to the ruby sdk which throws raise an error, so I will change it to throw an exception in the sync client, and complete exceptionally the future in the async client.